framework/include/utils/SerialAccess.h

```diff
@@ -0,0 +1,181 @@
+#define SERIAL_ACCESS_CONST_SIZE(type, dataptr, sizeval)                                           \
+  template <typename T>                                                                            \
+  struct SerialAccess<type<T>>                                                                     \
+  {                                                                                                \
+    static auto * data(type<T> & obj) { return dataptr; }                                          \
+    static constexpr std::size_t size(type<T> &) { return sizeval; }                               \
+    static constexpr std::size_t size() { return sizeval; }                                        \
+  }
+
+SERIAL_ACCESS_CONST_SIZE(VectorValue, &obj(0u), Moose::dim);
+SERIAL_ACCESS_CONST_SIZE(const VectorValue, &obj(0u), Moose::dim);
+SERIAL_ACCESS_CONST_SIZE(RankTwoTensorTempl, &obj(0u, 0u), RankTwoTensorTempl<T>::N2);
+SERIAL_ACCESS_CONST_SIZE(const RankTwoTensorTempl, &obj(0u, 0u), RankTwoTensorTempl<T>::N2);
+SERIAL_ACCESS_CONST_SIZE(RankFourTensorTempl, &obj(0u, 0u, 0u, 0u), RankFourTensorTempl<T>::N4);
+SERIAL_ACCESS_CONST_SIZE(const RankFourTensorTempl,
+                         &obj(0u, 0u, 0u, 0u),
+                         RankFourTensorTempl<T>::N4);
+
+// dynamic size containers (determining size requires an object instance)
+#define SERIAL_ACCESS_DYNAMIC_SIZE(type, dataptr, sizeval)                                         \
+  template <typename T>                                                                            \
+  struct SerialAccess<type<T>>                                                                     \
+  {                                                                                                \
+    static auto * data(type<T> & obj) { return dataptr; }                                          \
+    static constexpr std::size_t size(type<T> & obj) { return sizeval; }                           \
+  }
+
+SERIAL_ACCESS_DYNAMIC_SIZE(DenseVector, &obj(0u), obj.size());
+
+/**
+ * Value type helper (necessary for any type that does not have a value_type
+ * member or where value_type doesn't have a suitable meaning (ADReal)).
+ */
+template <typename T>
+struct SerialAccessVlaueTypeHelper
+{
+  typedef typename T::value_type value_type;
+};
+template <>
+struct SerialAccessVlaueTypeHelper<ADReal>
+{
+  typedef ADReal value_type;
+};
+template <>
+struct SerialAccessVlaueTypeHelper<Real>
+{
+  typedef Real value_type;
+};
+
+template <typename T>
+class SerialAccessRange
+{
+  typedef typename SerialAccessVlaueTypeHelper<typename std::remove_const<T>::type>::value_type R;
+  typedef typename std::conditional<std::is_const_v<T>, const R, R>::type V;
+
+public:
+  class iterator
+  {
+  public:
+    iterator(V * i) : _i(i) {}
+
+    V & operator*() const { return *_i; }
+
+    const iterator & operator++()
+    {
+      ++_i;
+      return *this;
+    }
+
+    iterator operator++(int)
```

Comment:
today I learned how we distinguish between pre-increment and post-increment operators 😄 There are pros to doing reviews for half my days...